scylla-server-rust/src/main.rs

```diff
@@ -7,44 +7,153 @@ use scylla_server_rust::{
 
 #[tokio::main]
 async fn main() {
-    let client: Database = Arc::new(PrismaClient::_builder().build().await.unwrap());
+    println!("Initializing scylla server...");
+    // construct a subscriber that prints formatted traces to stdout
+    // if RUST_LOG is not set, defaults to loglevel INFO
+    let subscriber = tracing_subscriber::fmt()
+        .pretty()
+        .with_thread_ids(true)
+        .with_ansi(true)
+        .with_thread_names(true)
+        .with_span_events(FmtSpan::CLOSE)
+        .with_env_filter(
+            EnvFilter::builder()
+                .with_default_directive(LevelFilter::INFO.into())
+                .from_env_lossy(),
+        )
+        .finish();
+    // use that subscriber to process traces emitted after this point
+    tracing::subscriber::set_global_default(subscriber).expect("Could not init tracing");
+
+    // create the database stuff
+    let db: Database = Arc::new(
+        PrismaClient::_builder()
+            .build()
+            .await
+            .expect("Could not build prisma DB"),
+    );
+
+    // create the socket stuff
+    let (socket_layer, io) = SocketIo::new_layer();
+    io.ns("/", |s: SocketRef| {
+        s.on_disconnect(|_: SocketRef| debug!("Socket: Client disconnected from socket"))
+    });
+
+    // channel to pass the mqtt data
+    // TODO tune buffer size
+    let (tx, rx) = mpsc::channel::<ClientData>(10000);
+
+    // channel to pass the processed data to the db thread
+    // TODO tune buffer size
+    let (tx_proc, rx_proc) = mpsc::channel::<Vec<ClientData>>(1000);
+
+    // the below two threads need to cancel cleanly to ensure all queued messages are sent.  therefore they are part of the a task tracker group.
+    // create a task tracker and cancellation token
+    let task_tracker = TaskTracker::new();
+    let token = CancellationToken::new();
+    // spawn the database handler
+    task_tracker.spawn(
+        db_handler::DbHandler::new(rx, Arc::clone(&db)).handling_loop(tx_proc, token.clone()),
+    );
+    // spawn the database inserter
+    task_tracker.spawn(db_handler::DbHandler::batching_loop(
+        rx_proc,
+        Arc::clone(&db),
+        token.clone(),
+    ));
+
+    // if PROD_SCYLLA=false
+    if std::env::var("PROD_SCYLLA").is_ok_and(|f| f == "false") {
+        info!("Running processor in mock mode, no data will be stored");
+        let recv = MockProcessor::new(io);
+        tokio::spawn(recv.generate_mock());
+    } else {
+        // run prod if this isnt present
+        // create and spawn the mqtt processor
+        info!("Running processor in MQTT (production) mode");
+        let (recv, eloop) = MqttProcessor::new(
+            tx,
+            std::env::var("PROD_SIREN_HOST_URL").unwrap_or("localhost:1883".to_string()),
+            db.clone(),
+            io,
+        )
+        .await;
+        tokio::spawn(recv.process_mqtt(eloop));
+    }
 
     let app = Router::new()
```

Comment:
this still isnt making new routers.

Reply:
I still don't understand the benefits to make a new router per 😢 .  Basically a new router in axum would not inherit state and CORS, meaning I need a seperate database reference and CORS layer for each router, resulting in a huge amount of ugliness and whatnot.  The closest I can get to what you are describing is https://github.com/tokio-rs/axum/issues/1326#issue-1350670719 which explains to create and clone an app state then merge them together.

Reply:
^^ I agree with Jack, I don't see the benefit when Scylla isn't going to have nearly as many routes as something like FinishLine (at least not soon). It feels like the technical lift would outweigh the negligible organizational benefit